Xamarin.Essentials/SaveToGallery/SaveToGallery.ios.macos.cs

```diff
@@ -0,0 +1,142 @@
+﻿using System;
+using System.IO;
+using System.Threading.Tasks;
+using Foundation;
+using Photos;
+
+namespace Xamarin.Essentials
+{
+    public static partial class SaveToGallery
+    {
+        static readonly string cacheDir = "EssentialsPhotosCacheDir";
+
+        static async Task PlatformSaveAsync(MediaFileType type, byte[] data, string fileName, string albumName)
+        {
+            string filePath = null;
+
+            try
+            {
+                filePath = GetFilePath(fileName);
+                await File.WriteAllBytesAsync(filePath, data);
+
+                await PlatformSaveAsync(type, filePath, albumName);
+            }
+            finally
+            {
+                DeleteFile(filePath);
+            }
+        }
+
+        static async Task PlatformSaveAsync(MediaFileType type, Stream fileStream, string fileName, string albumName)
+        {
+            string filePath = null;
+
+            try
+            {
+                filePath = GetFilePath(fileName);
+                using var stream = File.Create(filePath);
+                fileStream.CopyTo(stream);
+                stream.Close();
+
+                await PlatformSaveAsync(type, filePath, albumName);
+            }
+            finally
+            {
+                DeleteFile(filePath);
+            }
+        }
+
+        static async Task PlatformSaveAsync(MediaFileType type, string filePath, string albumName)
+        {
+            if (!File.Exists(filePath))
+                throw new ArgumentNullException(nameof(filePath));
+
+            await Permissions.EnsureGrantedAsync<Permissions.Photos>();
+
+            using var album = GetAlbum(albumName) ?? await CreateAlbumAsync(albumName);
```

Comment:
This part will fail on iOS 14 if limited access to the Media Library has been granted. See https://github.com/xamarin/Essentials/issues/1635 for details.

Basically, iOS 14 introduces a new access level to the gallery, allowing you to read selected photos and save new photos, but not create or retrieve named albums.

Using `await Permissions.EnsureGrantedAsync<Permissions.Photos>();` will not be enough, as it will pass but the code will later crash with a runtime exception missing permission to read or create the album.

Reply:
Yes, I know about the problem with iOS. I don't have a good idea for this yet